### PR TITLE
perf(neo4j): improve neo4j query performance by using node labels

### PR DESCRIPTION
**Pull request by pashashaik-mms**
**PR created and contributed by:** MediamarktSaturn Technology GmbH, Analytics-Services Team. Special thanks to @raudzis for the finding and idea proposed.

**PR Introduction:**
This PR introduces an optimization to the Neo4j querying process within our Datahub project. Previously, our Neo4j queries did not specify node labels during the match phase, which resulted in scanning all nodes in the database. This approach was inefficient, especially for large datasets. By integrating dynamic node labels into our match queries, we significantly improve query performance by leveraging Neo4j's ability to use indexes more effectively.

- **Node Label Integration:** Modified the Neo4j queries wherever applicable and now, the query explicitly targets nodes with the specified label, reducing the search space and improving performance.

- **Performance:** By applying node labels directly in our match clauses, the database engine can optimize node lookups using existing indexes, thus speeding up the query execution by reducing the number of nodes scanned.

- **Scalability:** These improvements make our database queries more scalable, handling larger datasets more efficiently.

- **Maintainability:** This change also enhances the clarity of our queries, making them more understandable at a glance, which benefits new contributors and maintainers alike.
-----------------------------------------------------------------------
## Checklist

- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- ~[ ] Links to related issues (if applicable)~
- [X] Tests for the changes have been added/updated (if applicable)
- ~[ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.~
- ~[ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)~

### CODE REVIEW

**Comment by pashashaik-mms** [comment]
This PR introduces an optimization to the Neo4j querying process within our Datahub project. Previously, our Neo4j queries did not specify node labels during the match phase, which resulted in scanning all nodes in the database. This approach was inefficient, especially for large datasets. By integrating dynamic node labels into our match queries, we significantly improve query performance by leveraging Neo4j's ability to use indexes more effectively.

**Node Label Integration:** Modified the Neo4j queries wherever applicable and now, the query explicitly targets nodes with the specified label, reducing the search space and improving performance.

**Performance:** By applying node labels directly in our match clauses, the database engine can optimize node lookups using existing indexes, thus speeding up the query execution by reducing the number of nodes scanned.
**Scalability:** These improvements make our database queries more scalable, handling larger datasets more efficiently.
**Maintainability:** This change also enhances the clarity of our queries, making them more understandable at a glance, which benefits new contributors and maintainers alike.

You can find the screenshot attached comparing the profile result:

**Profile WITHOUT NODELABEL query**
image

**Profile WITH NODELABEL query**
image

**Review by david-leifker — APPROVED** [review]

**Review by RyanHolstien — APPROVED** [review]

**Review by david-leifker — CHANGES_REQUESTED** [review]
There are broken tests with `./gradlew :metadata-io:test`

**Comment by pankajmahato-visa** [comment]
@pashashaik-mms 
 Run './gradlew :metadata-io:spotlessApply' to fix these violations.

**Comment by pashashaik-mms** [comment]
@RyanHolstien @david-leifker Could you please approve it as I had fixed the changes. It was a formatting issue and hence the build was failing and so are the unit-tests. I fixed it now and would need your approval.

**Comment by pashashaik-mms** [comment]
> Could you please approve it as I had fixed the changes. It was a formatting issue and hence the build was failing and so are the unit-tests. I fixed it now and would need your approval.

**Comment by deepgarg-visa** [comment]
@pashashaik-mms , as mentions [here](https://github.com/datahub-project/datahub/actions/runs/9198346655/job/25300909269?pr=10415#step:6:3427)

the below error is occurring because in testcases [here](https://github.com/datahub-project/datahub/blob/master/metadata-io/src/test/java/com/linkedin/metadata/graph/GraphServiceTestBase.java#L1108), "sourceEntityFilter" is passed as null and because of that the variable "srcNodeLabel" in method "findRelatedEntities" is not setting having a default value of blank, which results in the below query which is not correct.

`MATCH (src: )-[r:DownstreamOf ]-(dest ) WHERE left(type(r), 2)<>'r_'  RETURN dest, type(r) SKIP $offset LIMIT $count"`

Please also handle the case where variable "sourceEntityFilter" in method "findRelatedEntities" can be null or empty

`metadata-io > nonsearch > com.linkedin.metadata.graph.dgraph.DgraphGraphServiceTest > testFindRelatedEntitiesDestinationType[11](dataset, [HasOwner], {or=[], direction=UNDIRECTED}, [RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset1,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset2,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset3,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset4,PROD), via=null)]) STANDARD_ERROR
    org.neo4j.bolt.protocol.common.fsm.error.TransactionStateTransitionException: Invalid input ')': expected "%", "(" or an identifier (line 1, column 13 (offset: 12))
    "MATCH (src: )-[r:DownstreamOf ]-(dest ) WHERE left(type(r), 2)<>'r_'  RETURN dest, type(r) SKIP $offset LIMIT $count"`

**Comment by pashashaik-mms** [comment]
> @pashashaik-mms , as mentions [here](https://github.com/datahub-project/datahub/actions/runs/9198346655/job/25300909269?pr=10415#step:6:3427)
> 
> the below error is occurring because in testcases [here](https://github.com/datahub-project/datahub/blob/master/metadata-io/src/test/java/com/linkedin/metadata/graph/GraphServiceTestBase.java#L1108), "sourceEntityFilter" is passed as null and because of that the variable "srcNodeLabel" in method "findRelatedEntities" is not setting having a default value of blank, which results in the below query which is not correct.
> 
> `MATCH (src: )-[r:DownstreamOf ]-(dest ) WHERE left(type(r), 2)<>'r_' RETURN dest, type(r) SKIP $offset LIMIT $count"`
> 
> Please also handle the case where variable "sourceEntityFilter" in method "findRelatedEntities" can be null or empty
> 
> `metadata-io > nonsearch > com.linkedin.metadata.graph.dgraph.DgraphGraphServiceTest > testFindRelatedEntitiesDestinationType[11](dataset, [HasOwner], {or=[], direction=UNDIRECTED}, [RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset1,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset2,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset3,PROD), via=null), RelatedEntity(relationshipType=HasOwner, urn=urn:li:dataset:(urn:li:dataPlatform:type,SampleDataset4,PROD), via=null)]) STANDARD_ERROR org.neo4j.bolt.protocol.common.fsm.error.TransactionStateTransitionException: Invalid input ')': expected "%", "(" or an identifier (line 1, column 13 (offset: 12)) "MATCH (src: )-[r:DownstreamOf ]-(dest ) WHERE left(type(r), 2)<>'r_' RETURN dest, type(r) SKIP $offset LIMIT $count"`

FIXED NOW

**Comment by pashashaik-mms** [comment]
@deepgarg-visa fixed now. Could you please check and approve it. Its been waiting a while now.

**Comment by deepgarg-visa** [comment]
@pashashaik-mms are all metadata-io testcase passed ?

File: metadata-io/src/main/java/com/linkedin/metadata/graph/neo4j/Neo4jGraphService.java
Comment (deepgarg-visa) [inline metadata-io/src/main/java/com/linkedin/metadata/graph/neo4j/Neo4jGraphService.java:540]: this is common in both if and else. It can be put outside

File: metadata-io/src/main/java/com/linkedin/metadata/graph/neo4j/Neo4jGraphService.java
Comment (pashashaik-mms) [reply metadata-io/src/main/java/com/linkedin/metadata/graph/neo4j/Neo4jGraphService.java]: After a short vacation, I could see that you had already these changes. I made it in sync. What should be considered for this PR? Should we close it? or keep it for approval?

**Comment by david-leifker** [comment]
Looks like the changes were added in a different PR, closing this one. Thanks!

**Pull request closed without merging**
